Проектная работа №2

RD2W opened the pull request:
**Описание задания**

Приложение «Заметки» должно иметь возможность:
- Добавлять архивы (наборы заметок);
- Добавлять заметки в архивы;
- Просматривать заметки.

Важно, чтобы пользователь мог в любой момент выйти с экрана выбора архива, заметок или заметки и попасть на предыдущий список.

Итого получится пять экранов:
1. Выбор архива.
2. Создание архива.
3. Выбор заметки.
4. Создания заметки.
5. Экран заметки. На экране заметки должна быть возможность вывода текста заметки.

Эти пять экранов условно разделим на две группы:
1. Выбор элемента из списка и создание объекта.
2. Выбор архива, выбор заметки, экран заметки — это меню выбора.

**Важно**: выход из программы есть только один. Во всех остальных случаях мы возвращаемся на предыдущий экран.

Выбор пункта меню на каждом экране должен корректно реагировать на неправильный ввод:
- Если человек ввёл не цифру, то программа должна сказать ему, что следует вводить цифру, и ещё раз показать пункты меню.
- Если человек ввёл цифру, но такого элемента нет на экране, то программа должна сказать ему, что следует вводить цифру, которая соответствует одному из доступных пунктов меню , и ещё раз показать пункты меню.

**Критерии успешного выполнения задания:**

- Есть меню с возможностью добавления и просмотра архивов.
- Есть меню с возможностью добавления и просмотра заметок.
- Есть возможность добавлять и просматривать текст заметок.
- Не должно быть повторений одного и того же кода. Вся логика по считыванию ввода пользователя и вывода пунктов на экран должна переиспользовать один и тот же код.
- Ошибочный ввод пользователя должен корректно обрабатываться.
- Приложение не позволяет создавать архив или заметку без имени (с пустым именем).
- Приложение не позволяет создавать заметку без содержания (с пустым текстом).
- Из любого меню можно выйти и попасть на предыдущее меню или выйти из программы, если это просмотр архива.
- Приложение успешно компилируется и выполняется без ошибок.
- Весь код не написан в одном файле Main.
- Весь код приложения написан на Kotlin.

RD2W committed:
Создать дата класс Note

Создан дата класс для создания заметок, в котором для удобства переопределен метод toString().

RD2W committed:
Создать дата класс Archive

Создан дата класс для хранения заметок, в котором для удобства переопределен метод toString().

RD2W committed:
Создать класс Navigator

Создан класс для навигации по меню приложения и выполнения запросов от пользователя, для этой цели реализован дженерик метод navigate().

RD2W committed:
Создать класс ConsoleApp

Создан класс для отображения "экранов" приложения и взаимодействия с пользователем, для этой цели реализованы методы launch() и input().

RD2W committed:
Вызвать в main() требуемые методы для запуска приложения

RD2W committed:
Убрать лишний переход на новую строку при вызове println()

RD2W committed:
Убрать лишний переход на новую строку при вызове println() в main()

RD2W committed:
Переиспользовать метод input() в navigate() класса Navigator

RD2W committed:
Изменить UX для экрана открытой заметки

Переработана логика взаимодействия с пользователем. Данные изменения направлены на улучшение пользовательсткого опыта посредством изменения выпоняемых действий при открытии заметки. А именно: при открытии заметки сразу отображается её название, дата создания и содержимое.

RD2W committed:
Исправить опечатки в словах "Заметки" и "соответствующего"

i-masloed commented on src/main/kotlin/Archive.kt:
```diff
@@ -0,0 +1,5 @@
+data class Archive(val header: String, val notes: MutableList<Note>) {
+    override fun toString(): String {
```
Не рекомендуется переопределять так toString, лучше создать отдельный метод

RD2W replied:
Понял, в будущем буду учитывать данное замечание.

i-masloed commented on src/main/kotlin/ConsoleApp.kt:
```diff
@@ -0,0 +1,38 @@
+        var userInput: String
+        do {
+            userInput = readlnOrNull() ?: ""
+            val checkInput = userInput.isEmpty() || userInput.isBlank()
```
можно просто isBlank()

RD2W replied:
Исправлено, благодарю за подсказку!

RD2W committed:
Убрать избыточное условие для проверки ввода из консоли (не проверять на isEmpty())

@i-masloed, в очередной раз спасибо за подсказку!

RD2W committed:
Трансформировать класс Navigator в object Navigator

RD2W committed:
Изменить подход обработки ввода из консоли при навигации в меню программы

Рефакторин кода и уход от проверки на "регулярках", а именно использования matches(Regex("[0-${list.size + 1}]")).
